cmd/host/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+
+    core.halt()?;
+    core.read_8(var.addr, buf.as_mut_slice())?;
+    core.run()?;
+
+    Ok(buf)
+}
+
+fn host_boot_fail(hubris: &HubrisArchive, core: &mut dyn Core) -> Result<()> {
+    let d = read_var(hubris, core, "LAST_HOST_BOOT_FAIL")?;
+    if d.iter().all(|&c| c == 0) {
+        println!("[0; {}]", d.len());
+    } else {
+        match std::str::from_utf8(&d) {
+            Ok(s) => println!("{}", s.trim_matches('\0')),
+            Err(e) => println!("{:?}\n  (could not decode: {e})", d),
+        }
+    }
+
+    Ok(())
+}
+
+// Values and structs are defined in usr/src/uts/oxide/sys/kernel_ipcc.h
+// in the `stlouis` branch of `oxidecomputer/illumos-gate`
+const IPCC_PANIC_STACKS: usize = 0x10;
+const IPCC_PANIC_DATALEN: usize = 0x100;
+const IPCC_PANIC_SYMLEN: usize = 0x20;
+const IPCC_PANIC_MSGLEN: usize = 0x80;
+
+#[derive(Copy, Clone, Debug, FromBytes)]
+#[repr(packed)]
+struct IpccPanicData {
+    ipd_version: u8,
+    ipd_cause: u16,
+    ipd_error: u32,
+
+    ipd_cpuid: u32,
+    ipd_thread: u64,
+    ipd_addr: u64,
+    ipd_pc: u64,
+    ipd_fp: u64,
+    ipd_rp: u64,
+
+    ipd_message: [u8; IPCC_PANIC_MSGLEN],
+
+    ipd_stackidx: u8,
+    ipd_stack: [IpccPanicStack; IPCC_PANIC_STACKS],
+
+    ipd_dataidx: u8,
+    ipd_data: [u8; IPCC_PANIC_DATALEN],
+}
+
+#[derive(Copy, Clone, Debug, FromBytes)]
+#[repr(packed)]
+struct IpccPanicStack {
+    ips_symbol: [u8; IPCC_PANIC_SYMLEN],
+    ips_addr: u64,
+    ips_offset: u64,
+}
+
+fn host_last_panic(hubris: &HubrisArchive, core: &mut dyn Core) -> Result<()> {
+    let d = read_var(hubris, core, "LAST_HOST_PANIC")?;
+
+    // Fix for https://github.com/oxidecomputer/hubris/issues/1554
+    //
+    // In some cases, `ipd_cause` is unambiguous based on the first byte;
+    // otherwise, we populate a generic value.
+    let bonus_byte = match d[0] {
+        0x01 => None,
+        0xca => Some(0x11),
+        0x5e => Some(0x00),
+        0xa9 => Some(0x00),
+        0xeb => Some(0xff), // can't distinguish between different 0xeb**
+        b => bail!("could not decode `ipd_cause`: {b:#04x}"),
+    };
+    let d = if let Some(b) = bonus_byte {
+        msg!("fixing up trimmed initial data");
+        let mut next = vec![0x01, b];
+        next.extend(d);
+        next
+    } else {
+        d
+    };
+
+    let p = IpccPanicData::read_from_prefix(d.as_ref())
+        .context("failed to deserialize panic data")?;
+    if p.ipd_version != 0x01 {
+        bail!("unknown `ipd_version` ({})", p.ipd_version);
+    }
+    println!(
+        "ipd_cause:   {}",
+        match p.ipd_cause {
+            0xca11 => "IPCC_PANIC_CALL",
+            0xa900 => "IPCC_PANIC_TRAP",
+            0x5e00 => "IPCC_PANIC_USERTRAP",
+            0xeb00 => "IPCC_PANIC_EARLYBOOT",
+            0xeb97 => "IPCC_PANIC_EARLYBOOT_PROM",
+            0xeba9 => "IPCC_PANIC_EARLYBOOT_TRAP",
+            0xebff => "IPCC_PANIC_EARLYBOOT_*",
+            b => bail!("unknown `ipd_cause` {b:#04x}"),
```

Comment:
Tiny nit - since we already confirmed `ipd_version` above, maybe don't `bail!` here and just print this string instead? I think I'd expect additional causes being added would not cause a version number bump (if the structure of the data didn't change). Maybe the same argument applies to the `bonus_byte` workaround above, and that should similarly result in an "unknown cause" but still print the rest of the data.

Reply:
Sounds good, done in f5168a91fb7e69f3159efb0e15c1164850e1d31b